blockchain/validate.go

```diff
@@ -562,15 +569,25 @@ func checkBlockSanity(block *btcutil.Block, powLimit *big.Int, timeSource Median
 		// We could potentially overflow the accumulator so check for
 		// overflow.
 		lastSigOps := totalSigOps
-		totalSigOps += CountSigOps(tx)
-		if totalSigOps < lastSigOps || totalSigOps > MaxSigOpsPerBlock {
+		totalSigOps += (CountSigOps(tx) * WitnessScaleFactor)
+		if totalSigOps < lastSigOps || totalSigOps > MaxBlockSigOpsCost {
 			str := fmt.Sprintf("block contains too many signature "+
 				"operations - got %v, max %v", totalSigOps,
-				MaxSigOpsPerBlock)
+				MaxBlockSigOpsCost)
 			return ruleError(ErrTooManySigOps, str)
 		}
 	}
 
+	// Once the witness commitment, witness nonce, and sig op cost have
+	// been validated, we can finally assert that the block's weight
+	// doesn't exceed the current consensus paramter.
+	blockWeight := GetBlockWeight(block)
```

Comment:
Hmm, not sure I agree. With segwit, weight can actually entirely replace block size. Meaning that we could remove all instances of `MaxBlockSize` from the codebase and derive any limits related to max size from the witness scaling factor and other related constants. Block weight needs to be checked unconditionally, not only in the case that segwit has activated or not. Also stepping back and looking at the rest of the checks within the function, I'd say the check for the block weight nicely resides here. 

As a side note, leaving the block weight check in this function allows outside callers to take advantage of the non-contextual checks. For example, `neutrino` (our new light client) uses this function and therefore is able to properly validate the max block weight consensus limit (and the other relevant limits in this func) without an initialized instance of the `BlockChain` struct.